src/build/_util.py

```diff
@@ -48,6 +58,12 @@ def check_dependency(
             # dependency is satisfied.
             return
 
+    # Front ends SHOULD check explicitly for requirement cycles, and
+    # terminate the build with an informative message if one is found.
+    # https://www.python.org/dev/peps/pep-0517/#build-requirements
+    if project_name is not None and _normalize(req.name) == _normalize(project_name):
```

Comment:
You should perform this check on line 48; the current check is deficient.  It should be:

```py
    req = packaging.requirements.Requirement(req_string)
    canonicalized_name = packaging.utils.canonicalize_name(req.name)

[...]

    if any(
        packaging.utils.canonicalize_name(packaging.requirements.Requirement(a).name) == canonicalized_name
        for a in ancestral_req_strings
    ):
        raise CircularBuildDependencyError(...)

```

Reply:
> the current check is deficient

I'm having trouble getting this working, can you clarify what condition this is intended to catch.

I added a bunch of additional tests to ensure we are catching the various different dependency cycles.

Reply:
Apologies, this is fine as it is.